### PR TITLE
Update CNI metrics

### PR DESCRIPTION
**Pull request by mogren**
*Resolves issue #113 and #409 *

**Description of changes:**

* Replacing https://github.com/aws/amazon-vpc-cni-k8s/pull/88 with an up to date PR. (To see diff against #88, look at https://github.com/aws/amazon-vpc-cni-k8s/commit/c8faee22a000ca735ae4a69322e11534e753619e)
* Renamed prometheus endpoints to match the updated names in the CNI
* Made unit tests actually run
* Updated Makefile to include metrics target
* Made a dockerized metrics build
* Added `awscni_add_ip_req_count` and `awscni_del_ip_req_count`
* Made metrics agent version number match CNI release
* Made `USE_CLOUDWATCH="yes"` the default
* Passes go vet
* If the `CLUSTER_ID` tag is not set on the node, try `Name` before defaulting to `k8s-cluster`

**Examples**
To apply against CNI v1.4.0:
```
kubectl apply -f https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/release-1.4/config/v1.4/cni_metrics_helper.yaml
```
Check logs:
```
> kubectl get pods --all-namespaces  
NAMESPACE     NAME                                 READY   STATUS    RESTARTS   AGE
kube-system   aws-node-q64fb                       1/1     Running   0          4h
kube-system   cni-metrics-helper-9f9d65f97-h5tq5   1/1     Running   0          2h
kube-system   coredns-7d77776957-lx6fr             1/1     Running   0          17d
kube-system   kube-proxy-rvbkh                     1/1     Running   0          17d
```
```
> kubectl logs -n kube-system cni-metrics-helper-9f9d65f97-h5tq5
```
```
.
.
I0425 21:28:34.234286       1 metrics.go:329] Produce COUNTER metrics: awsAPIErr, value: 5.000000
I0425 21:28:34.234295       1 metrics.go:339] Produce GAUGE metrics: ipamdActionInProgress, value: 0.000000
I0425 21:28:34.234301       1 metrics.go:339] Produce GAUGE metrics: assignIPAddresses, value: 3.000000
I0425 21:28:34.234307       1 metrics.go:339] Produce GAUGE metrics: maxIPAddresses, value: 45.000000
I0425 21:28:34.234315       1 metrics.go:329] Produce COUNTER metrics: addReqCount, value: 161.000000
I0425 21:28:34.234324       1 metrics.go:329] Produce COUNTER metrics: delReqCount, value: 190.000000
I0425 21:28:34.234331       1 metrics.go:339] Produce GAUGE metrics: eniAllocated, value: 4.000000
I0425 21:28:34.234335       1 metrics.go:339] Produce GAUGE metrics: eniMaxAvailable, value: 9.000000
I0425 21:28:34.234341       1 metrics.go:339] Produce GAUGE metrics: totalIPAddresses, value: 20.000000
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cni-metrics-helper/cni-metrics-helper.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package main
```
Comment (vsiddharth) [inline cni-metrics-helper/cni-metrics-helper.go]: Missing license headers

File: cni-metrics-helper/metrics/cni_metrics.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
+// Package metrics handles the processing of all metrics. This file handles metrics for kube-state-metrics
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/cni_metrics.go]: Add license headers

File: cni-metrics-helper/metrics/cni_metrics.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+				matchFunc:  matchAny,
+				actionFunc: metricsAdd,
+				data:       &dataPoints{},
+				logToFile:  true}}},
+}
+
+// CNIMetricsTarget defines data structure for kube-state-metric target
+type CNIMetricsTarget struct {
+	interestingMetrics  map[string]metricsConvert
+	cwClient            publisher.Publisher
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/cni_metrics.go]: nit: s/cwClient/cwMetricsPublisher/

File: cni-metrics-helper/metrics/cni_metrics.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		cwClient:            cw,
+		kubeClient:          c,
+		discoveryController: d,
+		submitCW:            submitCW,
+	}
+}
+
+func (t *CNIMetricsTarget) grabMetricsFromTarget(cniPod string) ([]byte, error) {
+
+	glog.Info("Grabbing metrics from CNI ", cniPod)
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/cni_metrics.go]: nit: s/Info/Infof

File: cni-metrics-helper/metrics/cni_metrics.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	output, err := getMetricsFromPod(t.kubeClient, cniPod, metav1.NamespaceSystem, 61678)
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/cni_metrics.go]: Make 61678 a const

File: cni-metrics-helper/metrics/cni_metrics.go [hunk 6]
```diff
@@ -12,78 +12,92 @@ import (
 
 // InterestingCNIMetrics defines metrics parsing definition for kube-state-metrics
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/cni_metrics.go]: nit: Metrics from cni prometheus endpoint

File: cni-metrics-helper/metrics/metrics.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Package metrics provide common data structure and routines for converting/aggregating  prometheus metrics to cloudwatch metrics
+// Package metrics provide common data structure and routines for converting/aggregating prometheus metrics to cloudwatch metrics
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/metrics.go]: nit: Add license headers

File: cni-metrics-helper/metrics/metrics.go [hunk 8]
```diff
@@ -327,7 +318,7 @@ func filterMetrics(originalMetrics map[string]*dto.MetricFamily,
 }
 
 func produceCloudWatchMetrics(t metricsTarget, families map[string]*dto.MetricFamily, convertDef map[string]metricsConvert, cw publisher.Publisher) {
-	// TODO the following will be changed to integate with cloudwatch publisher
+	// TODO: The following will be changed to integrate with CloudWatch publisher
```
Comment (vsiddharth) [inline cni-metrics-helper/metrics/metrics.go]: nit: remove comment

File: pkg/ec2wrapper/ec2wrapper.go [hunk 9]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/golang/glog"
```
Comment (vsiddharth) [inline pkg/ec2wrapper/ec2wrapper.go]: Is glog used within this pkg?

File: pkg/publisher/publisher.go [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Package publisher is used to batch and send metric data to cloudwatch
+// Package publisher is used to batch and send metric data to CloudWatch
```
Comment (vsiddharth) [inline pkg/publisher/publisher.go]: nit: Add license headers

File: pkg/ec2wrapper/ec2wrapper.go (hunk 9 above)
Comment (mogren) [reply pkg/ec2wrapper/ec2wrapper.go]: Yes

**Pull request merged**